### PR TITLE
Node Request Certificates require to have IPs

### PR DESCRIPTION
**Pull request by aojea**
/kind bug

#### What this PR does / why we need it:

Before #121028 the hostname was always added to the Node object by the kubelet as address type `v1.NodeHostName`.

Post #121028 the kubelet waits for the external cloud provider to set all the Node addresses, `v1.NodeHostName` included.

#121028  caused an issue with some scenarios where the users want to override the hostname, and was fixed by #124516.

Post #124516 the kubelet initialize the Node object with the obtained hostname during the bootstrap, it can be overridden later by the cloud provider (unit test asserting this behavior in https://github.com/kubernetes/kubernetes/pull/125809)

Since now the Node object already contains at least one IP or DNSName, depending of the value of the hostname, the kubelet, when using the certificate manager, always requests a certificate at bootstrap with incomplete information, as it still may have to wait for the external cloud provider to populate the node addresses.

Make the condition to request the kubelet certificate to have at least one IP addresses, instead of having one IP or DNSName, solve the problem of kubelet requesting certificates too early.

An alternative is to wait for the external cloud provider taint to be removed

```release-note
kubelet now requests serving certificates only once it has at least one IP address in the `.status.addresses` of its associated Node object. This avoids requesting DNS-only serving certificates before externally set addresses are in place. Until 1.33, the previous behavior can be opted back into by setting the deprecated AllowDNSOnlyNodeCSR feature gate to true in the kubelet.
```

/sig node
/sig auth

/assign @danwinship @liggitt

### CODE REVIEW

File: pkg/kubelet/certificate/kubelet.go [hunk 1]
```diff
@@ -94,8 +94,13 @@ func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg
...
 		hostnames, ips := addressesToHostnamesAndIPs(getAddresses())
-		// don't return a template if we have no addresses to request for
-		if len(hostnames) == 0 && len(ips) == 0 {
+		// don't return a template if we have no addresses to request for.
+		// A Kubernetes Node requires to have at minimum one IP address
+		// because those are used on the Pods field HostIPs and in some cases,
+		// when pods uses hostNetwork: true, as PodIPs.
+		// Nodes that use IP addresses as Hostname are interpreted as an IP
+		// address, so it is possible that are nodes that don't have any DNS name.
+		if len(ips) == 0 {
```
Comment (liggitt) [inline pkg/kubelet/certificate/kubelet.go:103]: I'm not sure about this bit. DNS-only entries *could* be valid

File: pkg/kubelet/certificate/kubelet.go [hunk 2]
```diff
@@ -94,8 +94,13 @@ func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg
 
 	getTemplate := func() *x509.CertificateRequest {
 		hostnames, ips := addressesToHostnamesAndIPs(getAddresses())
-		// don't return a template if we have no addresses to request for
-		if len(hostnames) == 0 && len(ips) == 0 {
+		// don't return a template if we have no addresses to request for.
+		// A Kubernetes Node requires to have at minimum one IP address
+		// because those are used on the Pods field HostIPs and in some cases,
```
Comment (liggitt) [inline pkg/kubelet/certificate/kubelet.go:99]: it uses the IP from node.status.addresses as pod hostIP?

File: pkg/kubelet/certificate/kubelet.go (hunk 1 above)
Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:103]: it can be valid, but Kubernetes has a strong requirement on have at least one IP, as Pods depend on them, so my thinking is that a Node will need always at least one IP ... hence, any state without IPs  on the node.status.addresses is a transitional state

File: pkg/kubelet/certificate/kubelet.go (hunk 2 above)
Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:99]: yes, see https://docs.google.com/document/d/1mqdVLQHIYjrzjy8Hq-FMysthHRL7ht0guk2uWI0FxKM/edit#heading=h.97ewrhxwu05m fore more detailed explanation

**Comment by sftim** [comment]
Changelog suggestion
```diff
-kubelet, when configured to rotate certificates, will only request certificates when there is at least one IP address on the node.status.addresses
+Changed the way the kubelet requests certificates during rotation.
+The kubelet now only requests signing for a new certificate when there is at least one IP address in the
+`.status.addresses` of its associated Node object.
```

File: pkg/kubelet/certificate/kubelet.go (hunk 1 above)
Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:103]: the point of discussion here is when is really valid to request a certificate, do we want to request a certificate for a node that does not have any IP address reported?

Comment (sftim) [reply pkg/kubelet/certificate/kubelet.go:103]: I think maybe, because reporting ≠ having. Maybe the node is somehow multihomed but any connection based on a resolved IP address will stay up and working through ~magic~ technology.

Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:103]: This is a specific problem with external cloud providers that when the initialize the node object the set only the hostname, after that the cloud provider process the node and updates with the IP addresses.
The problem is that , if the node apporver does not consider valid a node with only DNS names and not with IP, there is a impact on node startup because nodes generate a first certificate request that is incomplete

**Comment by danwinship** [comment]
> An alternative is to wait for the external cloud provider to be removed

I assume the word "taint" is missing there? (Or are we removing the external cloud providers now that we've gotten rid of the legacy ones :joy:)

PR makes sense to me.

**Comment by aojea** [comment]
> > An alternative is to wait for the external cloud provider to be removed
> 
> I assume the word "taint" is missing there? (Or are we removing the external cloud providers now that we've gotten rid of the legacy ones 😂)
> 
> PR makes sense to me.

lol, yes, taint is indeed missing

File: pkg/kubelet/certificate/kubelet.go (hunk 2 above)
Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:99]: api docs https://github.com/kubernetes/kubernetes/blob/2ad04a0505ef13915e11d8b2b3445e514065a9bb/staging/src/k8s.io/api/core/v1/types.go#L6148-L6164

**Comment by liggitt** [comment]
I'm tentatively ok with this, given the way kubelet uses / requires an address of type IP for pod hostIP to work properly.

My read of kubelet address self-reporting is that in non-cloud envs, the kubelet always reports an IP (either detected or overridden with --node-ip), so this is scoped to external cloud providers.

A couple thoughts on reducing risk of this interacting weirdly with someone's setup:
1. check with folks that run kubelets in various envs or setup tools (e.g. @deads2k, @danwinship, @enj, @dims, @neolit123, @BenTheElder) that they don't know of weird setups this would cause issues with, for example:
   * kubelet serving expected to be working and secured using only the node DNS name in the critical path *before* the node IP address is assigned
   * the external cloud provider doesn't assign a node address of type IP at all (I think this would break pod hostIP functionality, so it seems really unlikely)
2. should we add a temporary gate (e.g. RequireIPKubeletServerCertificate, defaulting true) to allow opting out of this as a mitigation for a couple releases in case someone has a weird setup we didn't consider?

**Review by neolit123 — COMMENTED** [review]
the change seems ok to me, but is there a case where the cloud provider would only set an IP address in the status? if so, there should be a test case for that too.

seems like NewKubeletServerCertificateManager is not unit tested in terms of this hostname and IP validation.

**Comment by liggitt** [comment]
> is there a case where the cloud provider would only set an IP address in the status

only an IP is fine... that would have worked before, that would work with this PR as well (but it's good to have a unit test for it)

**Comment by liggitt** [comment]
> i can imagine scenarios like that - i.e. someone using the hostname to access the kubelet server and never using the IP. (IIUC this part)

If they report a DNS name, the kubelet will still include it in cert requests, and they can still use it as before. What would not work is if they *only* told the kubelet its DNS name, told it it had 0 IP addresses, and still expected kubelet to request serving certificates with just the DNS names and no IPs.

**Comment by neolit123** [comment]
>>> kubelet serving expected to be working and secured using only the node DNS name in the critical path before the node IP address is assigned

> > i can imagine scenarios like that - i.e. someone using the hostname to access the kubelet server and never using the IP. (IIUC this part)
> 
> If they report a DNS name, the kubelet will still include it in cert requests, and they can still use it as before. What would not work is if they _only_ told the kubelet its DNS name, told it it had 0 IP addresses, and still expected kubelet to request serving certificates with just the DNS names and no IPs.

i did not respond cleanly. as per your original comment they might assume they can use only hostname before there is a node ip, but that would fail because the CSR will not be created yet. so, yes i can image scenarios like that.

then again i'd say that's more of a corner case. in a CP scenario i would assume the controller adding IPs to nodes relatively fast. feature gating this would grant complaining users the toggle.

**Comment by liggitt** [comment]
> in a CP scenario i would assume the controller adding IPs to nodes relatively fast

Yeah, I'm less concerned with a delay and more concerned with a hard block, where the thing that will report the node IPs is hard blocked on something successfully using the kubelet serving endpoints secured by DNS name only. I'm not aware of anything like that, but wanted to explicitly ask folks that run kubelets in various envs, and give a short-term escape hatch for a couple releases in case someone in the wild is doing that.

**Comment by BenTheElder** [comment]
> check with folks that run kubelets in various envs or setup tools (e.g. @deads2k, @danwinship (acked https://github.com/kubernetes/kubernetes/pull/125813#issuecomment-2200318421, but just double-checking), @enj, @dims, @neolit123, @BenTheElder) that they don't know of weird setups this would cause issues with, for example:
> kubelet serving expected to be working and secured using only the node DNS name in the critical path before the node IP address is assigned

Have not seen this, not sure why it would be valid ... 


> the external cloud provider doesn't assign a node address of type IP at all (I think this would break pod hostIP functionality, so it seems really unlikely)

I haven't seen kubelet without node IP, only either assigned from kubelet or the cloud provider.

A lot of things in the ecosystem would break with no-IP nodes ... I really doubt that any cloud provider is not setting node IP, or that kubelet is otherwise being used without node IP ... but it's hard to prove I guess.

How long was this even possible for?

> Yeah, I'm less concerned with a delay and more concerned with a hard block, where the thing that will report the node IPs is hard blocked on something successfully using the kubelet serving endpoints secured by DNS name only. I'm not aware of anything like that, but wanted to explicitly ask folks that run kubelets in various envs, and give a short-term escape hatch for a couple releases in case someone in the wild is doing that.

This sounds reasonable.

**Comment by aojea** [comment]
added a feature gate to be able to opt-in for the old behavior

**Comment by aojea** [comment]
> Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] volumes should store data expand_more

/test pull-kubernetes-e2e-gce

File: pkg/kubelet/certificate/kubelet.go [hunk 3]
```diff
@@ -90,13 +92,28 @@ func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg
 			StabilityLevel: compbasemetrics.ALPHA,
 		},
 	)
-	legacyregistry.MustRegister(certificateRotationAge)
+	registerMetricsOnce.Do(func() {
```
Comment (liggitt) [inline pkg/kubelet/certificate/kubelet.go:95]: hmm, if newCertificateManagerConfig is called more than once, the returned configs will have metrics that aren't actually registered or reporting anywhere

before, MustRegister would panic on this... now, we just silently return not-working configs

if we're going to let this be called multiple times, I think the certificateRenewFailure / certificateRotationAge have to become shared package vars that lazily init / register once on first call

File: pkg/kubelet/certificate/kubelet.go [hunk 4]
```diff
@@ -90,13 +92,28 @@ func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg
...
+		legacyregistry.MustRegister(certificateRenewFailure)
+		legacyregistry.MustRegister(certificateRotationAge)
+	})
 
 	getTemplate := func() *x509.CertificateRequest {
 		hostnames, ips := addressesToHostnamesAndIPs(getAddresses())
-		// don't return a template if we have no addresses to request for
-		if len(hostnames) == 0 && len(ips) == 0 {
-			return nil
+		// don't return a template if we have no IP addresses or DNS names to request for
```
Comment (liggitt) [inline pkg/kubelet/certificate/kubelet.go:102]: I don't know why, but I'm finding the combination of the nesting and the feature gate name really hard to follow. Would something like this be clearer?

```go
// by default, require at least one IP before requesting a serving certificate
hasRequiredAddresses := len(ips) > 0

// optionally allow requesting a serving certificate with just a DNS name
if utilfeature.DefaultFeatureGate.Enabled(features.AllowDNSOnlyNodeCSR) {
  hasRequiredAddresses = hasRequiredAddresses || len(hostnames) > 0
}

// don't return a template if we have no addresses to request for
if !hasRequiredAddresses {
  return nil
}
```

Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:102]: is hard to follow but cleaner when we remove the feature gate, the diff is simpler

Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:102]: is hard to follow but it is cleaner, so when we remove the feature gate, the diff is simpler

Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:102]: will update with your suggestion

File: pkg/kubelet/certificate/kubelet.go (hunk 3 above)
Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:95]: > the returned configs will have metrics that aren't actually registered or reporting anywhere

metrics are global and you can only register once, so if is called multiple times the metrics will still apply

> before, MustRegister would panic on this... now, we just silently return not-working configs

why will be not-working? 

> if we're going to let this be called multiple times, I think the certificateRenewFailure / certificateRotationAge have to become shared package vars that lazily init / register once on first call

I started with that but the diff was larger( thinking on backports)

Comment (aojea) [reply pkg/kubelet/certificate/kubelet.go:95]: hmm, I didn't realize there is another certficiate.Config below 🤔

**Comment by aojea** [comment]
@liggitt the config generator was the wrong abstraction as it left the code worse than is today.
I took another approach and just decoupled the getTemplate() function to make it unit testable and code is more compatible with previous state

**Comment by aojea** [comment]
> TestCustomResourceDefaultingWithoutWatchCache

does not look related

/test pull-kubernetes-integration

**Comment by neolit123** [comment]
> > TestCustomResourceDefaultingWithoutWatchCache
> 
> does not look related
> 
> /test pull-kubernetes-integration

https://github.com/kubernetes/kubernetes/issues/125760

**Review by neolit123 — COMMENTED** [review]
/kind feature
/lgtm
